### PR TITLE
New "Alarm status" WS command type

### PR DESCRIPTION
**Pull request by dashevchenko**
## Pull Request description

A new ws command type was added to fetch the presence of active originator alarms. Command returns 'true' if at least one active alarm is present and 'false' otherwise.
PE: https://github.com/thingsboard/thingsboard-pe/pull/2994

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.
  
## Front-End feature checklist

- [ ] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [ ] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [ ] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [ ] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [ ] If new dependency was added: the dependency tree is checked for conflicts.
- [ ] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [ ] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [ ] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

File: application/src/main/java/org/thingsboard/server/service/ws/telemetry/cmd/v2/AlarmStatusUpdate.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import lombok.Getter;
+import lombok.ToString;
+import org.thingsboard.server.service.subscription.SubscriptionErrorCode;
+
+@ToString
+@Getter
+public class AlarmStatusUpdate extends CmdUpdate {
+
+    @Getter
+    private boolean present;
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/ws/telemetry/cmd/v2/AlarmStatusUpdate.java:29]: Let's rename to `active`

File: dao/src/main/java/org/thingsboard/server/dao/sql/alarm/AlarmRepository.java [hunk 2]
```diff
@@ -283,10 +273,8 @@ Page<AlarmInfoEntity> findCustomerAlarms(@Param("tenantId") UUID tenantId,
                     "WHERE a.tenantId = :tenantId AND a.customerId = :customerId " +
                     "AND (:startTime IS NULL OR a.createdTime >= :startTime) " +
                     "AND (:endTime IS NULL OR a.createdTime <= :endTime) " +
-                    "AND ((:#{#alarmTypes == null} = true) OR a.type IN (:alarmTypes)) " + //HHH-15968
-                    "AND ((:#{#alarmSeverities == null} = true) OR a.severity IN (:alarmSeverities)) " + //HHH-15968
-//                    "AND ((:alarmTypes) IS NULL OR a.type IN (:alarmTypes)) " +
-//                    "AND ((:alarmSeverities) IS NULL OR a.severity IN (:alarmSeverities)) " +
+                    "AND ((:alarmTypes) IS NULL OR a.type IN (:alarmTypes)) " +
+                    "AND ((:alarmSeverities) IS NULL OR a.severity IN (:alarmSeverities)) " +
```
Comment (ViacheslavKlimov) [inline dao/src/main/java/org/thingsboard/server/dao/sql/alarm/AlarmRepository.java:277]: Why is this changed? Is HHH-15968 no longer an issue?

File: dao/src/main/java/org/thingsboard/server/dao/sql/alarm/AlarmRepository.java [hunk 3]
```diff
@@ -404,4 +392,25 @@ String updateAlarm(@Param("t_id") UUID tenantId, @Param("a_id") UUID alarmId, @P
...
     int deleteTypeIfNoAlarmsExist(@Param("tenantId") UUID tenantId, @Param("types") Set<String> types);
 
+    @Query(value = "SELECT a.id " +
+            "FROM AlarmEntity a " +
+            "WHERE a.tenantId = :tenantId " +
+            "AND a.originatorId = :originatorId " +
+            "AND ((:alarmTypes) IS NULL OR a.type IN (:alarmTypes)) " +
+            "AND ((:alarmSeverities) IS NULL OR a.severity IN (:alarmSeverities)) " +
+            "AND (a.cleared = false)",
+            countQuery = "" +
```
Comment (ViacheslavKlimov) [inline dao/src/main/java/org/thingsboard/server/dao/sql/alarm/AlarmRepository.java:402]: Do we use count? Using Page will cause Spring Data to perform two queries - one for data and one for count.
Let's use `Slice` to avoid this (see findAlarmIdsByAssigneeId in this class)

File: dao/src/main/java/org/thingsboard/server/dao/sql/alarm/AlarmRepository.java [hunk 4]
```diff
@@ -404,4 +392,25 @@ String updateAlarm(@Param("t_id") UUID tenantId, @Param("a_id") UUID alarmId, @P
     @Query(value = "DELETE FROM alarm_types AS at WHERE NOT EXISTS (SELECT 1 FROM alarm AS a WHERE a.tenant_id = at.tenant_id AND a.type = at.type) AND at.tenant_id = :tenantId AND at.type IN (:types)", nativeQuery = true)
     int deleteTypeIfNoAlarmsExist(@Param("tenantId") UUID tenantId, @Param("types") Set<String> types);
 
+    @Query(value = "SELECT a.id " +
```
Comment (ViacheslavKlimov) [inline dao/src/main/java/org/thingsboard/server/dao/sql/alarm/AlarmRepository.java:395]: Please attach the result of `EXPLAIN ANALYZE` for this query. Should we include tenantId in the search?

File: application/src/main/java/org/thingsboard/server/service/subscription/TbAlarmStatusSubscription.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    @Override
+    public boolean equals(Object o) {
+        return super.equals(o);
+    }
+
+    @Override
+    public int hashCode() {
+        return super.hashCode();
+    }
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/subscription/TbAlarmStatusSubscription.java:63]: No need to override if you call `super`

File: application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbEntityDataSubscriptionService.java [hunk 6]
```diff
@@ -527,7 +607,7 @@ private TbAlarmDataSubCtx createSubCtx(WebSocketSessionRef sessionRef, AlarmData
         return ctx;
     }
 
-    private TbAlarmCountSubCtx createSubCtx(WebSocketSessionRef sessionRef, AlarmCountCmd cmd) {
+    private  TbAlarmCountSubCtx createSubCtx(WebSocketSessionRef sessionRef, AlarmCountCmd cmd) {
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbEntityDataSubscriptionService.java:610]: Fix the formatting back :)

File: application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbEntityDataSubscriptionService.java [hunk 7]
```diff
@@ -434,6 +444,76 @@ public void handleCmd(WebSocketSessionRef session, AlarmCountCmd cmd) {
...
+
+    private void handleAlarmStatusSubscriptionUpdate(TbSubscription<AlarmSubscriptionUpdate> sub, AlarmSubscriptionUpdate subscriptionUpdate) {
+        TbAlarmStatusSubscription subscription = (TbAlarmStatusSubscription) sub;
+        try {
+            AlarmInfo alarm = subscriptionUpdate.getAlarm();
+            Set<UUID> alarmsIds = subscription.getAlarmIds();
+            if (alarmsIds.contains(alarm.getId().getId())) {
+                if (!alarmMatchesSubscription(alarm, subscription) || subscriptionUpdate.isAlarmDeleted()) {
+                    alarmsIds.remove(alarm.getId().getId());
+                    if (alarmsIds.size() == 0) {
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbEntityDataSubscriptionService.java:490]: -> `alarmsIds.isEmpty()`

File: application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbEntityDataSubscriptionService.java [hunk 8]
```diff
@@ -434,6 +444,76 @@ public void handleCmd(WebSocketSessionRef session, AlarmCountCmd cmd) {
...
+        sendUpdate(sessionRef.getSessionId(), subscription.createUpdate());
+    }
+
+    private void fetchActiveAlarms(TbAlarmStatusSubscription subscription) {
+        log.trace("[{}, subId: {}] Fetching active alarms from DB", subscription.getSessionId(), subscription.getSubscriptionId());
+        OriginatorAlarmFilter originatorAlarmFilter = new OriginatorAlarmFilter(subscription.getEntityId(), subscription.getTypeList(), subscription.getSeverityList());
+        List<UUID> alarmIds = alarmService.findActiveOriginatorAlarms(subscription.getTenantId(), originatorAlarmFilter, new PageLink(alarmsPerAlarmStatusSubscriptionCacheSize)).getData();
+
+        subscription.getAlarmIds().addAll(alarmIds);
+        subscription.setExceededLimit(alarmIds.size() == alarmsPerAlarmStatusSubscriptionCacheSize);
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbEntityDataSubscriptionService.java:474]: Let's rename to `setCacheFull`

File: application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbEntityDataSubscriptionService.java [hunk 9]
```diff
@@ -434,6 +444,76 @@ public void handleCmd(WebSocketSessionRef session, AlarmCountCmd cmd) {
...
+                    sendUpdate(subscription.getSessionId(), subscription.createUpdate());
+                }
+            }
+        } catch (Exception e) {
+            log.error("[{}, subId: {}] Failed to handle update for alarm status subscription: {}", subscription.getSessionId(), subscription.getSubscriptionId(), subscriptionUpdate, e);
+        }
+    }
+
+    private boolean alarmMatchesSubscription(AlarmInfo alarm, TbAlarmStatusSubscription subscription) {
+        return !alarm.isCleared() && (subscription.getTypeList() == null || subscription.getTypeList().contains(alarm.getType())) &&
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbEntityDataSubscriptionService.java:513]: Let's move to TbAlarmStatusSubscription: `matches(Alarm)`

File: dao/src/main/java/org/thingsboard/server/dao/sql/alarm/AlarmRepository.java (hunk 2 above)
Comment (dashevchenko) [reply dao/src/main/java/org/thingsboard/server/dao/sql/alarm/AlarmRepository.java:277]: yes it is fixed and new jpa version is being used in the project

File: dao/src/main/java/org/thingsboard/server/dao/sql/alarm/AlarmRepository.java [hunk 10]
```diff
@@ -392,25 +404,14 @@ String updateAlarm(@Param("t_id") UUID tenantId, @Param("a_id") UUID alarmId, @P
...
-                    "AND ((:alarmTypes) IS NULL OR a.type IN (:alarmTypes)) " +
-                    "AND ((:alarmSeverities) IS NULL OR a.severity IN (:alarmSeverities)) " +
-                    "AND (a.cleared = false)")
-    Page<UUID> findActiveOriginatorAlarms(@Param("tenantId") UUID tenantId,
-                                          @Param("originatorId") UUID originatorId,
+    @Query(value = "SELECT a.id FROM alarm a " +
+            "WHERE a.originator_id = :originatorId " +
+            "AND (COALESCE(:alarmTypes) IS NULL OR a.type IN (:alarmTypes)) " +
+            "AND (COALESCE(:alarmSeverities) IS NULL OR a.severity IN (:alarmSeverities)) " +
+            "AND (a.cleared = false) LIMIT :limit", nativeQuery = true)
```
Comment (ViacheslavKlimov) [inline dao/src/main/java/org/thingsboard/server/dao/sql/alarm/AlarmRepository.java:411]: Please add order by id

File: application/src/main/java/org/thingsboard/server/service/subscription/TbAlarmStatusSubscription.java [hunk 11]
```diff
@@ -37,7 +38,7 @@ public class TbAlarmStatusSubscription extends TbSubscription<AlarmSubscriptionU
     private final Set<UUID> alarmIds = new HashSet<>();
     @Getter
     @Setter
-    private boolean exceededLimit;
+    private boolean fullCache;
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/subscription/TbAlarmStatusSubscription.java:41]: I think "cacheFull" suits better
